plugins/net_plugin/net_plugin.cpp

```diff
@@ -2182,14 +2184,38 @@ namespace eosio {
               sync_next_expected_num < sync_last_requested_num );
    }
 
+   // called from c's connection strand
    bool sync_manager::is_sync_request_ahead_allowed(block_num_type blk_num) const REQUIRES(sync_mtx) {
```

Comment:
It doesn't wait till we have 0 blocks that can be applied before requesting more. But I reworked it to use your `num_blocks_not_applied` and `num_blocks_that_can_be_applied` variables to be clearer.

Reply:
> It doesn't wait till we have 0 blocks that can be applied before requesting more.

I'm pretty sure it did. Before if would not return true unless the following condition is true:

`if (calculated_lib <= head_num) {`

which occurs only when we have applied all the blocks possible.

Reply:
In the case of LIB stuck, yes. But in the nominal case it would return `true` via the `num_block_not_applied` `if`.

Reply:
correct, that's the case I was referring to. 

When lib is not stuck this second part is unnecessary anyways, since `num_blocks_not_applied` and `num_blocks_that_can_be_applied` are essentially the same (modulo the number of blocks that lib trials back for).